### ISSUE TITLE
Spacing has changed

### ISSUE THREAD

**Issue by Kethku** [dfrg/swash#16]
In Neovide we determine the grid size by looking at the average glyph width and assuming that for monospace fonts, that will give us a reasonable size for a character grid.

However in your recent change, that average width has increased. Plus I think thats actually a bad metric to use because some fonts have a few very wide glyphs, so it would be incorrect to use the average glyph width.

I'm wondering if you have a better recommendation for what metric to look at. I've seen some systems pick an arbitrary character such as capital X and assume that it is representative. I've also attempted in the past to actually shape a representative piece of text and measure the spacing used there are the ground truth for the entire grid.

To be honest I don't know what the "Correct" solution is. Does swash have any metric that would work well for this use case?

**Comment by dfrg** [comment]
I've seen the screenshot in the linked issue and that spacing is definitely odd. It looks like the shaper and renderer are using slightly different scaling factors. The aspect ratios of the individual glyphs seem off as well but that may just be Skia attempting to mimic macOS stem darkening.

The average width metric is pulled directly out of the OS/2 table without modification, so I'm not sure what might have changed on this end. 

As far as alternative metrics, CSS offers the ch unit which is the advance width of the digit 0. It also might be worth considering computing a new average width from a set of representative characters. Unfortunately, all options here are going to be heuristic. 

This really does look like an out of sync scaling factor though.

**Comment by dustinroepsch** [comment]
I just realized, this is might not be new behavior. 

By default, PragmataPro isn't a pure monospace fault, all of the text is monospace but glyphs can be wider. 
In my original screenshot, I was using PragmataProMonoLiga, which has mono glyphs the same size as normal text.
The weird spacing is from regular PragmataPro. Maybe PragmataPro reports the widest possible glyph as the font width, even though most characters are supposed to be smaller. 

If variable width fonts are out of scope of this project (and neovide), feel free to close this issue!
I'll just use the monospaced version of PragmataPro :)

**Comment by dfrg** [comment]
That is curious. So PragmataProMono yields reasonable spacing in Neovide even with the changes?

When you have a moment, it might be helpful to open both fonts at https://opentype.js.org/font-inspector.html Expand the "OS/2 and Windows metrics table" row and report the value of `xAvgCharWidth` for each font. 

If it's substantially larger for the non-mono font, it might be useful to have a check in Neovide to determine if the advance width of X is less than the average width by some threshold and then compute a new appropriate cell width as an average of some representative set of characters.

I'm not sure if Neovide wants to reject proportional fonts outright. I know VS Code refuses to render with fonts where `advance('i') != advance('w')`

**Comment by dustinroepsch** [comment]
Yep! PragmataProMono looks great, even with ligatures :)
<img width="699" alt="Screen Shot 2021-08-03 at 12 09 32 PM" src="https://user-images.githubusercontent.com/5170924/128072280-442eb83e-e243-494c-ac92-a1c279f0b768.png">
I'll report back with the font inspector results later tonight, gotta stop procrastinating my day job right now... :^)

**Comment by Kethku** [comment]
@dfrg something to note is that neovide doesn't currently use the advance provided by swash for calculating positioning of glyphs. This is because I couldn't depend on fonts to have consistent spacing themselves even when they are called "monospace". Instead the system uses the average glyph widget. So my hypothosis is that PragmataProMonoLiga has a different average glyph width because the ligatures often stretch past the single character size.

TLDR I don't think the spacing thing is an upstream issue, but rather a side effect of a hack I added

**Comment by dfrg** [comment]
> I'll report back with the font inspector results later tonight, gotta stop procrastinating my day job right now... :^)

No rush :) It's just mostly to confirm what we already suspect about the average width.

> @dfrg something to note is that neovide doesn't currently use the advance provided by swash for calculating positioning of glyphs. This is because I couldn't depend on fonts to have consistent spacing themselves even when they are called "monospace". Instead the system uses the average glyph widget.

This is a reasonable approach and has seemed to work well so far. A slightly more sophisticated heuristic might be able to accommodate ostensibly monospace fonts such as the one at issue here if you want to go that route.
